cpp/ql/src/Best Practices/Likely Errors/OffsetUseBeforeRangeCheck.ql

```diff
@@ -23,6 +24,7 @@ predicate beforeArrayAccess(Variable v, ArrayExpr access, Expr before) {
   )
 }
 
+bindingset[v, after]
```

Comment:
I'm not sure `bindingset`s is the right tool here. We're basically just inlining those predicates and hoping for a more lucky join order now. Instead, I think we should be more precise about which one we want to join first.

You mentioned that `beforeArrayAccess` and `afterArrayAccess` was joined in the wrong order. What was the wrong order exactly? I'd imagine that we want to join with `afterArrayAccess` first since that's the non-negated one in the query. Is that correct?

If that's correct, I'd suggest sticking an `only_bind_into` on _all_ the arguments to `afterArrayAccess`. This will force `afterArrayAccess` to be joined first. This is one of the most common ways of using `only_bind_into`. For more information on this, see [this internal link](https://github.com/github/codeql-team/blob/main/docs/best-practices/join-order.md#evaluate-p-first).

Reply:
It was `getArrayOffset()` exploding, with a load of context pulled in that I think came from the `from-where-select`.  So my goal was just to get these predicates evaluated with some sensible assumptions, but you're right, `bindingset` is not the right tool.

That said I'm not keen on carpet bombing it with `only_bind_into` unless we really need to (I admit it would probably be a robust solution).  So I've gone with `nomagic`, which I think expresses exactly what I wanted in the first place.

Reply:
Ah, that makes sense. One of the common optimizations done by the compiler is converting `a and not b` to `a and not (a and b)` to reduce the size of the negation. So I can imagine some bad context got pulled into `beforeArrayAccess`. If that's what happens, do we really need the `nomagic` on `afterArrayAccess`?

Reply:
We probably don't, but I think it makes sense to be consistent as `beforeArrayAccess` and `afterArrayAccess` are so similar.